File: 0038.Count_and_Say.py

#12.34%RT
class Solution(object):
  def countAndSay(self, n):
      s = '1'
      for _ in range(n - 1):
          s = re.sub(r'(.)\1*', lambda m: str(len(m.group(0))) + m.group(1), s)
      return s



#21.62%RT
class Solution(object):
  def countAndSay(self, n):
      s = '1'
      for _ in range(n - 1):
          s = ''.join(str(len(group)) + digit
                      for group, digit in re.findall(r'((.)\2*)', s))
      return s


#5.47%RT
class Solution(object):
  def countAndSay(self, n):
      s = '1'
      for _ in range(n - 1):
          s = ''.join(str(len(list(group))) + digit
                      for digit, group in itertools.groupby(s))
      return s

File: 0038.Count_and_Say.rs
//100%RT

use std::iter::once;

impl Solution {
    pub fn count_and_say(n: i32) -> String {
        let mut curr = vec![1];
        for _ in 1..n {
            let mut next = vec![];
            let mut slow = 0;
            for fast in 0..=curr.len() {
                if fast == curr.len() || curr[slow] != curr[fast] {
                    next.extend(once((fast - slow) as u8).chain(once(curr[slow] as u8)));
                    slow = fast;
                }
            }
            curr = next;
        }
        curr.into_iter().map(|digit| (digit + b'0') as char).collect()
    }
}
